File: toki/solutions/toki_dp.py
import functools
import sys

sys.setrecursionlimit(100000)

CONSONANTS, VOWELS, ILLEGALS = 'mnptkswjl', 'aeiou', 'wu wo ji ti nn nm'

def solve(W: str) -> str:
    """
    Uses top-down dynamic programming to check the syllable structure, after
    first checking for illegal sequences.
    """    
    for i in range(len(W) - 1):
        if W[i:i + 2] in ILLEGALS or W[i] in VOWELS and W[i + 1] in VOWELS:
            return 'ike'
    
    @functools.cache
    def dp(i):
        """
        Returns True if and only if W[i:] has a valid syllable structure.
        """
        if i == len(W):
            return True
        elif i > len(W):
            return False
        else:
            return is_valid_syllable(W[i:i + 1]) and dp(i + 1) or \
                   is_valid_syllable(W[i:i + 2]) and dp(i + 2) or \
                   is_valid_syllable(W[i:i + 3]) and dp(i + 3)
    return 'pona' if dp(0) else 'ike'

def is_valid_syllable(s):
    if len(s) == 1:
        return s in VOWELS
    elif len(s) == 2:
        return s[0] in CONSONANTS and s[1] in VOWELS or \
               s[0] in VOWELS and s[1] == 'n'
    elif len(s) == 3:
        return s[0] in CONSONANTS and s[1] in VOWELS and s[2] == 'n'
    else:
        return False

def main():
    T = int(input())
    for _ in range(T):
        W = input()
        print(solve(W))

if __name__ == '__main__':
    main()

File: toki/solutions/toki_dp.cpp
#include <iostream>
#include <vector>

using namespace std;

string CONSONANTS = "mnptkswjl", VOWELS = "aeiou";
string ILLEGALS = "wu wo ji ti nn nm";

#define CONTAINS(text, pattern) text.find(pattern) != string::npos

bool is_valid_syllable(string s);

/**
 * Uses bottom-up dynamic programming to check the syllable structure, after
 * first checking for illegal sequences.
 */
string solve(string W) {
    for (int i = 0; i < W.length() - 1; i++) {
        if (CONTAINS(ILLEGALS, W.substr(i, 2)) ||
            CONTAINS(VOWELS, W[i]) && CONTAINS(VOWELS, W[i + 1])) {
            return "ike";
        }
    }
    
    // dp[i] is true if and only if W.substr(i) has a valid syllable structure.
    bool dp[W.length() + 3];
    dp[W.length()] = true;
    dp[W.length() + 1] = dp[W.length() + 2] = false;
    for (int i = W.length() - 1; i >= 0; i--) {
        dp[i] = dp[i + 1] && is_valid_syllable(W.substr(i, 1)) ||
                dp[i + 2] && is_valid_syllable(W.substr(i, 2)) ||
                dp[i + 3] && is_valid_syllable(W.substr(i, 3));
    }
    
    if (dp[0]) {
        return "pona";
    } else {
        return "ike";
    }
}

bool is_valid_syllable(string s) {
    switch (s.length()) {
        case 1:
            return CONTAINS(VOWELS, s[0]);
        case 2:
            return CONTAINS(CONSONANTS, s[0]) && CONTAINS(VOWELS, s[1]) ||
                   CONTAINS(VOWELS, s[0]) && s[1] == 'n';
        case 3:
            return CONTAINS(CONSONANTS, s[0]) &&
                   CONTAINS(VOWELS, s[1]) &&
                   s[2] == 'n';
        default:
            return false;
    }
}

int main() {
    int T;
    cin >> T;
    for (int i = 0; i < T; i++) {
        string W;
        cin >> W;
        cout << solve(W) << '\n';
    }
}
